File: src/main/java/skywolf46/commandannotation/minecraft/MinecraftCommandImpl.java
package skywolf46.commandannotation.minecraft;

import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import skywolf46.commandannotation.data.autocomplete.AutoCompleteSupplier;
import skywolf46.commandannotation.data.command.CommandArgument;
import skywolf46.commandannotation.data.methodprocessor.MethodChain;
import skywolf46.commandannotation.util.ParameterStorage;

import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

public class MinecraftCommandImpl extends Command {
    private MethodChain chain;
    private HashMap<String, MinecraftCommandImpl> chainSubCommands = new HashMap<>();

    public MinecraftCommandImpl(String commandName) {
        super(commandName);
    }


    @Override
    public boolean execute(CommandSender commandSender, String s, String[] strings) {
        ParameterStorage storage = ParameterStorage.of(commandSender);
        long key = System.currentTimeMillis();
        CommandArgument arg = new CommandArgument(storage, s, strings, key);
        storage.set(arg);
        MinecraftCommandImpl implt = this;
        for (int i = 0; i < strings.length; i++) {
            if (implt.chainSubCommands.containsKey(strings[i])) {
                implt = implt.chainSubCommands.get(strings[i]);
                arg.nextPointer();
            }
        }
        implt.invokeCommand(storage);
        return true;
    }

    @Override
    public List<String> tabComplete(CommandSender sender, String alias, String[] args) throws IllegalArgumentException {
        MinecraftCommandImpl impl = this;
        long key = System.currentTimeMillis();
        ParameterStorage stor = new ParameterStorage();
        CommandArgument arg = new CommandArgument(stor, alias, args, key);

        int i = 0;
        for (; i < args.length; i++) {
            if (impl.chainSubCommands.containsKey(args[i])) {
                impl = impl.chainSubCommands.get(args[i]);
                arg.nextPointer();
            } else break;
        }
        stor.set(arg);
        stor.add(sender);
        List<String> completer = impl.chainSubCommands.keySet().stream().filter(r -> args.length == 0 || args[args.length - 1].isEmpty() || r.toLowerCase().startsWith(args[args.length - 1].toLowerCase())).collect(Collectors.toList());
        if (impl.chainSubCommands.size() <= 0 && args.length > 0) {
            for (Player pl : Bukkit.getOnlinePlayers()) {
                if (pl.getName().toLowerCase().startsWith(args[args.length - 1])) {
                    completer.add(pl.getName());
                }
            }
        }
        if (impl.chain != null) {
            AutoCompleteSupplier sup = impl.chain.getCompleteSupplier();
            if (sup != null) {
                sup.editCompletion(stor, completer);
            }
        }
        return completer;
    }

    private void invokeCommand(ParameterStorage storage) {
        if (chain == null)
            return;
        chain.invoke(storage);
    }

    public void insert(String[] n, MethodChain chain) {
        if (n.length == 0) {
            this.chain = chain;
        } else {
            MinecraftCommandImpl impl = this;
            for (int i = 1; i < n.length; i++) {
                impl = impl.chainSubCommands.computeIfAbsent(n[i], a -> {
                    return new MinecraftCommandImpl("");
                });
            }
            impl.chain = chain;
        }
    }

    public void invokeSubCommand(CommandArgument arg) {

    }
}

File: minecraft/src/main/kotlin/skywolf46/commandannotationmc/minecraft/impl/minecraft/MinecraftCommandImpl.kt
package skywolf46.commandannotationmc.minecraft.impl.minecraft

import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import skywolf46.commandannotation.kotlin.data.Arguments
import skywolf46.commandannotationmc.minecraft.CommandAnnotation
import skywolf46.extrautility.data.ArgumentStorage

class MinecraftCommandImpl(starting: String) : Command(starting) {
    override fun execute(p0: CommandSender, p1: String, p2: Array<String>): Boolean {
        val arguments = Arguments(false, "/$p1", ArgumentStorage(), p2, 0)
        arguments.addParameter(p0)
        arguments.addParameter(arguments._storage)
        arguments.addParameter(arguments)
        CommandAnnotation.command.inspect("/$p1", arguments).apply {
            // Initialize pointer
            arguments._sysPointer = first
            second.forEach {
                if (!it.invoke(arguments.clone()))
                    return false
            }
        }
        return true
    }

    override fun tabComplete(sender: CommandSender, alias: String, args: Array<String>): List<String> {
        val arg = Arguments(false, "/$alias", ArgumentStorage(), args, 0)
        arg.addParameter(sender)
        CommandAnnotation.command.inspect("/$alias", arg).apply {
            if (second.isEmpty() && second[0].getCompleter() != null)
                second[0].getCompleter()!!.findCompleter(arg)
        }
        return CommandAnnotation.command.inspectNextParameter("/$alias", arg)
    }
}